torchdynamo/variables/tensor.py

```diff
@@ -671,6 +676,17 @@ def from_tensor_variable(cls, tensor_variable, raw_value, need_unwrap=True):
             need_unwrap=need_unwrap,
         )
 
+    def convert_to_constant(self, tx):
+        for graph_arg in tx.output.graphargs:
+            if graph_arg.source is self.source:
+                graph_arg.erase()
+
+        for g in self.guards:
+            if g is self.volatile_guard:
+                g.create_fn = GuardBuilder.CONSTANT_MATCH
+
+        return ConstantVariable(value=self.raw_value, guards=self.guards)
```

Comment:
hmm - is this safe? do the old guards made for the last object apply here? I ask because VariableBuilder thinks that `CONSTANT_MATCH` suffices for a constant variable, maybe just have that one guard?

Reply:
L686 updated the match method, so it's safe.

Reply:
This is an extension of my previous question - why make >1 `constant_match`? why even have volatile guards? why not just make all new guards here? 

Maybe something much more explicit like:

```
self.guards.remove(type_match)
self.guards.remove(x)
return ConstantVariable(value=self.raw_value, guards=guards(source, name, CONSTANT_MATCH)
```

That way, we are much clearer, it stays simpler, and we do not need to introduce a new notion of volatile guards. 

Also... what about guards that are not create_fn assigned, like tensor_guards? will this still invoke `___check_tensors`?

Reply:
Still the same, the guards on this variable have been propagated to many down streams. If we do sth like you mentioned, we only update the guard on this variable. But if down stream variables depends on this variable's guard as well, it still uses the old guards(type_match), which cause errors. I think most of your comment here is about this guard replacement, I'm happy to schedule a VC chat to discuss if it can help.

Reply:
BTW, ```volatile_guard``` is used to help us find out which guard need to be replaced, because there are possible multiple guards on the same variable, we need to identify which is exactly the unspec guard.

Reply:
> BTW, volatile_guard is used to help us find out which guard need to be replaced, because there are possible multiple guards on the same variable, we need to identify which is exactly the unspec guard.

I understand what you are doing. I just think it is too complicated. 

You only have a single volatile guard I see you make here, from `GuardBuilder.TYPE_MATCH` - why do we need a whole new concept of `volatile_guard` vertically integrated through so many layers? Why not just find the `TYPE_MATCH` guard and remove it explicitly?

Reply:
> Why not just find the TYPE_MATCH guard and remove it explicitly?

That's exactly what we want to do. A variable can have ___multiple___ guards of ```TYPE_MATCH```, what we can know which guard is the one we want to replace?

Reply:
1) Would you not remove all the TYPE_MATCH in this case? When would you only remove some? 
2) If you do only need some, so even if you need partial removal, you don't need to make a new volatile_guard notion and store it in kwargs, or `self` - you can just use TYPE_MATCH itself, right? 

instead of:

```
for g in self.guards:
       if g is self.volatile_guard:
```

Can you store in on TYPE_MATCH?

```
for g in self.guards
    if g.is_volatile()
```

And then you are marking the guard itself, instead of having the object store its guard

Reply:
1, We don't remove all ```TYPE_MATCH```, only remove the one binding to the specific unspec variable.
2, ```TYPE_MATCH``` is a function, so we can't store it on ```TYPE_MATCH```.

Reply:
1) Understood
2) Why not?

```
def foo(x, y, z):
  return x + y + z

f = foo
f.volatile = True
print(f.volatile)
```

Reply:
Python is a ~~godless~~ dynamic and exciting language, you can even do:

```
def foo(x, y, z):
  return x + y + z

foo.foo = foo
foo.foo(1, 2, 3)
```

Reply:
It seems it works well, but if we can put it as an attribute of class/object, why instead we put it as an attribute of function? I think the former is easy to read and understand.

Reply:
A function is also an instance of an object here. The benefit here is that its cleaner - what's volatile? the guard is. Instead of parallel bookkeeping, just keep it on the fn or on the guard - this also means that you can handle volatile guards anywhere, you do not need to track the object itself - for this case its the same, but there may be future cases where you just want to iterate over or manage volatile guards without having a reference to their originating object. 

TLDR is, guard is volatile, so store volatile on the guard.

Reply:
Thank you for making the change!